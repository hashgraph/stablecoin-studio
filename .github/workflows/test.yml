name: test
on:
  pull_request:
    branches: [ main, release/**, feature/** ]
  push:
    branches: [ main, release/** ]
    tags: [ v* ]
jobs:
  test-node:
    name: Unit Tests
    runs-on: ubuntu-latest
    steps:
      - name: Use Node.js TLS 16.x
        uses: actions/setup-node@v1
        with:
          node-version: 16.x
      - name: Checkout repository
        uses: actions/checkout@v2
      - name: Install dependencies
        run: npm i
      - name: Build
        run: npm run build
      - name: Test
        run: npm run test