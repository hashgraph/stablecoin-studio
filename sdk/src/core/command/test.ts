/* eslint-disable @typescript-eslint/no-empty-interface */

interface CommandResponse {}
interface ICommand {}

export class Command<T> implements ICommand {
	private $resultType!: T;
}

export interface ICommandBus {
	execute<X>(command: Command<X>): Promise<X>;
}

export type IInferringCommandHandler<CommandType extends Command<unknown>> =
	CommandType extends Command<infer ResultType>
		? {
				execute(command: CommandType): Promise<ResultType>;
		  }
		: never;

class ConcreteCommandResponse implements CommandResponse {
	constructor(public readonly status: string) {}
}

class ConcreteCommand extends Command<ConcreteCommandResponse> {
	constructor(public readonly value: string) {
		super();
	}
}

class CommandHandler<CommandType extends Command<unknown>>
	implements IInferringCommandHandler<ConcreteCommand>
{
	execute(command: ConcreteCommand): Promise<ConcreteCommandResponse> {
		throw new Error('Method not implemented.');
	}
}

class CommandBus implements ICommandBus {
	execute<X>(command: Command<X>): Promise<X> {
		throw new Error('Method not implemented.');
	}
}

// Test

const bus = new CommandBus();
const res = bus.execute(new ConcreteCommand('1'));
