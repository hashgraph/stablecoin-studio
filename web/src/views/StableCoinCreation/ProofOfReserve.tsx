import { Heading, HStack, Stack, Text, VStack } from '@chakra-ui/react';
import type { CreateRequest } from 'hedera-stable-coin-sdk';
import { useEffect } from 'react';

import type { Control, FieldValues, UseFormReturn } from 'react-hook-form';
import { useWatch } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import InputController from '../../components/Form/InputController';
import SwitchController from '../../components/Form/SwitchController';
import { propertyNotFound } from '../../constant';
import { handleRequestValidation } from '../../utils/validationsHelper';

interface ProofOfReserveProps {
	form: UseFormReturn;
	control: Control<FieldValues>;
	request: CreateRequest;
}

const ProofOfReserve = (props: ProofOfReserveProps) => {
	const { control, form, request } = props;
	const { t } = useTranslation(['global', 'stableCoinCreation']);

	const proofOfReserve = useWatch({
		control,
		name: 'proofOfReserve',
	});

	const hasDataFeed = useWatch({
		control,
		name: 'hasDataFeed',
	});

	useEffect(() => {
		if (!proofOfReserve) {
			form.resetField('reserveInitialAmount');
			form.resetField('reserveAddress');
		}
	}, [proofOfReserve]);

	useEffect(() => {
		if (!hasDataFeed) {
			form.resetField('reserveAddress');
		} else {
			form.resetField('reserveInitialAmount');
		}
	}, [hasDataFeed]);

	return (
		<VStack h='full' justify={'space-between'} pt='80px'>
			<Stack minW={400}>
				<Heading
					data-testid='title'
					fontSize='16px'
					fontWeight='600'
					mb={10}
					lineHeight='15.2px'
					textAlign={'left'}
				>
					{t('stableCoinCreation:proofOfReserve.title')}
				</Heading>

				<HStack>
					<Text maxW={'252px'} fontSize='14px' fontWeight='400' lineHeight='17px'>
						{t('stableCoinCreation:proofOfReserve.description')}
					</Text>
					<SwitchController control={control} name={'proofOfReserve'} defaultValue={false} />
				</HStack>

				{proofOfReserve === true && (
					<HStack>
						<Text maxW={'252px'} fontSize='14px' fontWeight='400' lineHeight='17px'>
							{t('stableCoinCreation:proofOfReserve.haveDataFeed')}
						</Text>
						<SwitchController control={control} name={'hasDataFeed'} defaultValue={false} />
					</HStack>
				)}

				{hasDataFeed === true && proofOfReserve === true && (
					<HStack pt={'15px'}>
						<InputController
							rules={{
								required: t(`global:validations.required`) ?? propertyNotFound,
								validate: {
									validation: (value: string) => {
										request.createReserve = proofOfReserve;
										request.reserveAddress = value;
										const res = handleRequestValidation(request.validate('reserveAddress'));
										return res;
									},
								},
							}}
							isRequired
							control={control}
							name={'reserveAddress'}
							label={t('stableCoinCreation:proofOfReserve.dataFeed') ?? propertyNotFound}
							placeholder={
								t('stableCoinCreation:proofOfReserve.dataFeedPlaceholder') ?? propertyNotFound
							}
						/>
					</HStack>
				)}

				{!hasDataFeed && proofOfReserve === true && (
					<HStack pt={'15px'}>
						<InputController
							rules={{
								required: t(`global:validations.required`) ?? propertyNotFound,
								validate: {
									validation: (value: string) => {
										request.createReserve = proofOfReserve;
										request.reserveInitialAmount = value;
										const res = handleRequestValidation(request.validate('reserveInitialAmount'));
										return res;
									},
								},
							}}
							isRequired
							control={control}
							name={'reserveInitialAmount'}
							label={t('stableCoinCreation:proofOfReserve.initialSupply') ?? propertyNotFound}
							placeholder={t('stableCoinCreation:proofOfReserve.initialSupply') ?? propertyNotFound}
						/>
					</HStack>
				)}
			</Stack>
		</VStack>
	);
};

export default ProofOfReserve;
