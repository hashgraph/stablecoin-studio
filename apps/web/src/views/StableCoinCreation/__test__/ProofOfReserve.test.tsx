import { FieldValues, useForm } from 'react-hook-form';
import { render } from '../../../test/index';
import translations from '../../../translations/en/stableCoinCreation.json';
import ProofOfReserve from '../ProofOfReserve';
import { CreateRequest } from '@hashgraph/stablecoin-npm-sdk';

jest.mock('react-hook-form', () => ({
	...jest.requireActual('react-hook-form'),
	Controller: () => <></>,
	useWatch: () => jest.fn(),
	useForm: () => ({
		control: () => ({}),
		handleSubmit: () => jest.fn(),
		resetField: () => jest.fn(),
	}),
}));

const form = useForm<FieldValues>({
	mode: 'onChange',
});

const { control } = form;
const request = new CreateRequest({
	name: '',
	symbol: '',
	decimals: 6,
	createReserve: false,
});

describe(`<${ProofOfReserve.name} />`, () => {
	test('should render correctly', () => {
		const component = render(<ProofOfReserve control={control} request={request} form={form} />);

		expect(component.asFragment()).toMatchSnapshot();
	});

	test('should has title', () => {
		const component = render(<ProofOfReserve control={control} request={request} form={form} />);

		const header = component.getByTestId('title');
		expect(header).toHaveTextContent(translations.proofOfReserve.title);
	});
});
