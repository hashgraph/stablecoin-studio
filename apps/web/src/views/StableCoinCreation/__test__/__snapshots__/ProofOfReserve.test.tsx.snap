// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProofOfReserve /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1f6llri"
  >
    <div
      class="chakra-stack css-135a0es"
    >
      <h2
        class="chakra-heading css-p7jgmy"
        data-testid="title"
      >
        Proof-of-Reserve
      </h2>
      <div
        class="chakra-stack css-84zodg"
      >
        <p
          class="chakra-text css-3crm2z"
        >
          Will your stablecoin be collateralized by another on-chain asset?
        </p>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
